File: frontend/src/types/research.ts
export type ResearchMode = "environment" | "company" | "person" | "market" | "job" | "product";
export type DepthMode = "quick" | "standard" | "deep";

export interface UserProfile {
  user_id: number;
  email: string;
  display_name: string;
  created_at: string;
}

export interface AuthResponse {
  token: string;
  user: UserProfile;
}

export interface SessionSummary {
  session_id: string;
  title: string;
  mode: ResearchMode;
  depth: DepthMode;
  created_at: string;
  updated_at: string;
  pinned: boolean;
  bookmarked: boolean;
  compare_selected: boolean;
  tags: string[];
  latest_query?: string | null;
}

export interface ResearchMessage {
  message_id: string;
  role: "user" | "assistant";
  content: string;
  created_at: string;
}

export interface SourceItem {
  source_id: string;
  title: string;
  source_type: string;
  snippet: string;
  credibility: number;
  url?: string | null;
  cluster?: string | null;
  agreement: "supports" | "mixed" | "weak";
  metadata?: Record<string, unknown> | null;
}

export interface AgentStep {
  step_id: string;
  name: string;
  status: "pending" | "running" | "completed" | "failed" | "partial";
  summary: string;
  started_at: string;
  finished_at?: string | null;
  payload?: Record<string, unknown> | null;
}

export interface ConfidenceBreakdown {
  overall: number;
  data_quality: number;
  coverage: number;
  reasoning_strength: number;
}

export interface ClaimScore {
  claim: string;
  confidence: number;
  supporting_sources: string[];
  contradiction_sources: string[];
}

export interface ReportSection {
  key: string;
  title: string;
  content: string;
  sources: string[];
}

export interface ResearchReport {
  report_id: string;
  session_id: string;
  query: string;
  title: string;
  mode: ResearchMode;
  depth: DepthMode;
  overview: string;
  sections: ReportSection[];
  follow_ups: string[];
  sources: SourceItem[];
  claims: ClaimScore[];
  contradictions: string[];
  confidence: ConfidenceBreakdown;
  summary_markdown: string;
  created_at: string;
}

export interface SessionDetail {
  session: SessionSummary;
  messages: ResearchMessage[];
  reports: ResearchReport[];
  steps: AgentStep[];
}

export interface ResearchQueryResponse {
  session: SessionSummary;
  messages: ResearchMessage[];
  report: ResearchReport;
  steps: AgentStep[];
  degraded: boolean;
}

export interface ExportResponse {
  filename: string;
  format: "json" | "markdown" | "pdf";
  download_url: string;
}

export interface KnowledgeNode {
  node_id: string;
  label: string;
  group: "session" | "mode" | "tag" | "source";
}

export interface KnowledgeEdge {
  source: string;
  target: string;
  weight: number;
}

export interface KnowledgeGraphResponse {
  nodes: KnowledgeNode[];
  edges: KnowledgeEdge[];
}

export interface DashboardContext {
  site_id: number;
  station_label: string;
  selected_timestamp?: string | null;
  selected_label?: string | null;
  phase?: "live" | "forecast" | null;
  visible_o3: number;
  visible_no2: number;
  risk: string;
  horizon_hours: number;
  profile_focus: string;
  pollutant_focus: string;
  model_name: string;
  network_scope: "india" | "global";
  selected_city_key?: string | null;
  selected_city_name?: string | null;
  selected_city_label?: string | null;
  selected_city_o3?: number | null;
  selected_city_no2?: number | null;
  selected_city_aqi?: number | null;
  forecast_peak_o3: number;
  forecast_peak_no2: number;
  forecast_peak_o3_time: string;
  forecast_peak_no2_time: string;
  alert_count: number;
  top_alert_pollutant?: string | null;
  top_alert_severity?: string | null;
  baseline_avg_o3?: number | null;
  baseline_avg_no2?: number | null;
}

File: backend/app/schemas/research.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


ResearchMode = Literal["environment", "company", "person", "market", "job", "product"]
DepthMode = Literal["quick", "standard", "deep"]
StepStatus = Literal["pending", "running", "completed", "failed", "partial"]
ExportFormat = Literal["json", "markdown", "pdf"]


class SignUpRequest(BaseModel):
    email: str
    password: str = Field(min_length=8)
    display_name: str | None = None


class SignInRequest(BaseModel):
    email: str
    password: str


class UserProfile(BaseModel):
    user_id: int
    email: str
    display_name: str
    created_at: datetime


class AuthResponse(BaseModel):
    token: str
    user: UserProfile


class SourceItem(BaseModel):
    source_id: str
    title: str
    source_type: str
    snippet: str
    credibility: float
    url: str | None = None
    cluster: str | None = None
    agreement: Literal["supports", "mixed", "weak"] = "supports"
    metadata: dict | None = None


class AgentStep(BaseModel):
    step_id: str
    name: str
    status: StepStatus
    summary: str
    started_at: datetime
    finished_at: datetime | None = None
    payload: dict | None = None


class ConfidenceBreakdown(BaseModel):
    overall: float
    data_quality: float
    coverage: float
    reasoning_strength: float


class ClaimScore(BaseModel):
    claim: str
    confidence: float
    supporting_sources: list[str]
    contradiction_sources: list[str] = []


class ReportSection(BaseModel):
    key: str
    title: str
    content: str
    sources: list[str] = []


class ResearchReport(BaseModel):
    report_id: str
    session_id: str
    query: str
    title: str
    mode: ResearchMode
    depth: DepthMode
    overview: str
    sections: list[ReportSection]
    follow_ups: list[str]
    sources: list[SourceItem]
    claims: list[ClaimScore]
    contradictions: list[str]
    confidence: ConfidenceBreakdown
    summary_markdown: str
    created_at: datetime


class SessionSummary(BaseModel):
    session_id: str
    title: str
    mode: ResearchMode
    depth: DepthMode
    created_at: datetime
    updated_at: datetime
    pinned: bool = False
    bookmarked: bool = False
    compare_selected: bool = False
    tags: list[str] = []
    latest_query: str | None = None


class ResearchMessage(BaseModel):
    message_id: str
    role: Literal["user", "assistant"]
    content: str
    created_at: datetime


class SessionDetail(BaseModel):
    session: SessionSummary
    messages: list[ResearchMessage]
    reports: list[ResearchReport]
    steps: list[AgentStep]


class SessionCreateRequest(BaseModel):
    title: str | None = None
    mode: ResearchMode = "environment"
    depth: DepthMode = "standard"
    tags: list[str] = []


class SessionUpdateRequest(BaseModel):
    title: str | None = None
    pinned: bool | None = None
    bookmarked: bool | None = None
    compare_selected: bool | None = None
    tags: list[str] | None = None


class ResearchQueryRequest(BaseModel):
    session_id: str | None = None
    query: str
    mode: ResearchMode = "environment"
    depth: DepthMode = "standard"
    site_id: int | None = None
    network_scope: Literal["india", "global"] = "india"
    dashboard_context: dict[str, Any] | None = None


class ResearchQueryResponse(BaseModel):
    session: SessionSummary
    messages: list[ResearchMessage]
    report: ResearchReport
    steps: list[AgentStep]
    degraded: bool = False


class FeedbackRequest(BaseModel):
    report_id: str
    target_type: Literal["summary", "section", "source"]
    target_key: str
    value: Literal["positive", "negative"]
    notes: str | None = None


class ExportResponse(BaseModel):
    filename: str
    format: ExportFormat
    download_url: str


class KnowledgeNode(BaseModel):
    node_id: str
    label: str
    group: Literal["session", "mode", "tag", "source"]


class KnowledgeEdge(BaseModel):
    source: str
    target: str
    weight: float = 1.0


class KnowledgeGraphResponse(BaseModel):
    nodes: list[KnowledgeNode]
    edges: list[KnowledgeEdge]
